### ISSUE TITLE
MediaImplementation.TakeMediaAsync Exception

### ISSUE THREAD

**Issue by ericbrunner** [jamesmontemagno/MediaPlugin#118]
## Bug 

Sometimes on Android when taking a photo I get that excpetion.

Version Number of Plugin: 2.6.0.12-beta
Device Tested On: LENOVO T2-A7-20F

Package: com.rolsped.TruckerApp
Version Code: 27
Version Name: 1.0
Android: 4.4.2
Android Build: T2-A7-20F_S000009_151022_ROW
Manufacturer: Lenovo
Model: Tab2A7-20F
Thread: main-1
Start Date: Wed Oct 26 20:56:22 CEST 2016
Date: Wed Oct 26 21:03:27 CEST 2016
Format: Xamarin

Xamarin caused by: android.runtime.JavaProxyThrowable: System.InvalidOperationException: Only one operation can be active at a time
  at Plugin.Media.MediaImplementation.TakeMediaAsync (System.String type, System.String action, Plugin.Media.Abstractions.StoreMediaOptions options) [0x00039] in <35066fbd82df477487a6a1e39050b480>:0 
  at Plugin.Media.MediaImplementation+<PickPhotoAsync>d__17.MoveNext () [0x0008a] in <35066fbd82df477487a6a1e39050b480>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <fe5ae17f49cd42d9923d7a78735dd424>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <fe5ae17f49cd42d9923d7a78735dd424>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <fe5ae17f49cd42d9923d7a78735dd424>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <fe5ae17f49cd42d9923d7a78735dd424>:0 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <fe5ae17f49cd42d9923d7a78735dd424>:0 
  at trucker_rolsped.Pages.MediaPage+<<-ctor>b__5_2>d.MoveNext () [0x001ce] in <097cdfc3784648edab9e909af0f616aa>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <fe5ae17f49cd42d9923d7a78735dd424>:0 
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore.<ThrowAsync>m__0 (System.Object state) [0x00000] in <fe5ae17f49cd42d9923d7a78735dd424>:0 
  at Android.App.SyncContext+<Post>c__AnonStorey0.<>m__0 () [0x00000] in <79e6945a2f4a4e2f844e36c860dfb012>:0 
  at Java.Lang.Thread+RunnableImplementor.Run () [0x0000b] in <79e6945a2f4a4e2f844e36c860dfb012>:0 
  at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00009] in <79e6945a2f4a4e2f844e36c860dfb012>:0 
  at (wrapper dynamic-method) System.Object:93cfcaba-4c78-4a88-a4df-f54a276b6364 (intptr,intptr)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:30)
	at android.os.Handler.handleCallback(Handler.java:808)
	at android.os.Handler.dispatchMessage(Handler.java:103)
	at android.os.Looper.loop(Looper.java:193)
	at android.app.ActivityThread.main(ActivityThread.java:5341)
	at java.lang.reflect.Method.invokeNative(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:515)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:825)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:641)
	at dalvik.system.NativeStart.main(Native Method)

**Comment by sbonello** [comment]
Seems there are concurrent calls to TakeMediaAsync. have a look at this solution 

http://damianblog.com/2016/07/06/only-one-operation-can-be-active-at-at-time/

**Comment by ericbrunner** [comment]
@sbonello Thanks. My code is from the sample like that. The only difference is that I create a copy of the media stream because xamarin thankfully disposes the stream when loaded with ImageSource.FromStream and I need the stream for upload afterwards to the image preview.

    takePhoto.Clicked += async (sender, args) =>
            {
                await CrossMedia.Current.Initialize();

                if (!CrossMedia.Current.IsCameraAvailable || !CrossMedia.Current.IsTakePhotoSupported)
                {
                    await DisplayAlert("Media Info", "Keine Kamera verfügbar.", "OK");
                    return;
                }

                SelectedMedia = new TruckerAppMediaFile
                {
                    Type = MediaType.Photo,
                    File = await CrossMedia.Current.TakePhotoAsync(new Plugin.Media.Abstractions.StoreCameraMediaOptions
                    {
                        PhotoSize = PhotoSize.Medium,
                        CompressionQuality = 92,
                        SaveToAlbum = true,
                        Directory = "TruckerAppPhotos",
                        Name = $"{DateTime.Now}_documenttypeXYZ.jpg"
                    })
                };


                if (SelectedMedia?.File == null)
                {
                    await DisplayAlert("Media Info", "Es wurde kein Foto erstellt.", "OK");
                    return;
                }


                var imageStream = await DependencyService.Get<Interfaces.Media.IMedia>()
                    .CopyMediaAsync(SelectedMedia.File.GetStream());
                image.Source = ImageSource.FromStream(() => imageStream);

                SendMedia.IsEnabled = true;
            };

**Comment by sbonello** [comment]
If somehow multiple concurrent click events are generated the exception will be generated.

            var ntcs = new TaskCompletionSource<MediaFile>(id);
            if (Interlocked.CompareExchange(ref this.completionSource, ntcs, null) != null)
                throw new InvalidOperationException("Only one operation can be active at a time");

I would guard against this by adding a boolean check . The function will be async only on the first await 

        takePhoto.Clicked += async (sender, args) =>
        {
            if (_loaded)
                 return;

           _loaded = true;

            await CrossMedia.Current.Initialize();

            if (!CrossMedia.Current.IsCameraAvailable || !CrossMedia.Current.IsTakePhotoSupported)
            {
                await DisplayAlert("Media Info", "Keine Kamera verfügbar.", "OK");
                return;
            }

            SelectedMedia = new TruckerAppMediaFile
            {
                Type = MediaType.Photo,
                File = await CrossMedia.Current.TakePhotoAsync(new Plugin.Media.Abstractions.StoreCameraMediaOptions
                {
                    PhotoSize = PhotoSize.Medium,
                    CompressionQuality = 92,
                    SaveToAlbum = true,
                    Directory = "TruckerAppPhotos",
                    Name = $"{DateTime.Now}_documenttypeXYZ.jpg"
                })
            };
  

           _loaded = false;

**Comment by ericbrunner** [comment]
@sbonello Thank you very much. I will add that. The curious part that it happens when I click the Pick Image button and select the Google Fotos App.

**Comment by ericbrunner** [comment]
@sbonello I get that petmission exception when I pick a foto and select Google Fotos App:

android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> Java.Lang.SecurityException: Permission Denial: reading com.google.android.apps.photos.contentprovider.MediaContentProvider uri content://com.google.android.apps.photos.contentprovider/-1/1/content%3A%2F%2Fmedia%2Fexternal%2Fimages%2Fmedia%2F368/ORIGINAL/NONE/1886983672 from pid=3817, uid=10117 requires the provider be exported, or grantUriPermission()

**Comment by jamesmontemagno** [comment]
Will close  this out as it looks like you guys figured it out :) awesome

**Comment by ericbrunner** [comment]
@jamesmontemagno
Why closed? It is not solved.

**Comment by jamesmontemagno** [comment]
It is, you can't call the method twice without having the first finished.

**Comment by ericbrunner** [comment]
It is not! I dont call the method twice. I only call it once. Anyway I will catch the exception when I pick a foto from Google Fotos and post it here or in a new issue. 
I added yesterday the got permission exception. (see above)

**Comment by ericbrunner** [comment]
@jamesmontemagno On Device LENOVO Tablet Tab2A7-20F Android 4.4 API LEVEL 19 I sometimes get that exception when I pick an image from the latest **Google Fotos** App:

`android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> Java.Lang.SecurityException: Permission Denial: reading com.google.android.apps.photos.contentprovider.MediaContentProvider uri content://com.google.android.apps.photos.contentprovider/-1/1/content%3A%2F%2Fmedia%2Fexternal%2Fimages%2Fmedia%2F368/ORIGINAL/NONE/1886983672 from pid=3817, uid=10117 requires the provider be exported, or grantUriPermission()`

This is the debug info:

```
SelectedMedia?.File
{Plugin.Media.Abstractions.MediaFile}
    AlbumPath: "content://com.google.android.apps.photos.contentprovider/-1/1/content%3A%2F%2Fmedia%2Fexternal%2Ffil…"
    Path: "/storage/emulated/0/Android/data/com.rolsped.TruckerApp/files/Pictures/temp/IMG_20161027_173717.jpg"
SelectedMedia?.File.Path
"/storage/emulated/0/Android/data/com.rolsped.TruckerApp/files/Pictures/temp/IMG_20161027_173717.jpg"
SelectedMedia?.File.AlbumPath
"content://com.google.android.apps.photos.contentprovider/-1/1/content%3A%2F%2Fmedia%2Fexternal%2Ffile%2F366/ORIGINAL/NONE/1706517233"
```

**Comment by jamesmontemagno** [comment]
Yes, this is different then what was reported, so would be a different bug, so I suggest opening a new one and reporting what version of the plugin you are using with the info you put on there.
